lib/open_project/translations/models/locales_updater.rb

```diff
@@ -0,0 +1,186 @@
+# todo those ought to be in the gemspec
+require 'tempfile'
+require 'crowdin-api'
+require 'yaml'
+require 'zip'
+require 'pathname'
+require 'fileutils'
+
+require_relative '../helpers/tmp_directory'
+require_relative './git_repository'
+
+ENGLISH_TRANLATION_FILE = 'en.yml'
+ENGLISH_JS_TRANLATION_FILE = 'js-en.yml'
+
+class LocalesUpdater
+  extend TmpDirectory
+
+  def self.update_all_locales_of_all_repos(debug: true)
+    repos_to_update = plugins_with_locales
+
+    within_tmp_directory(delete_if_exists: true, debug: debug) do
+      repos_to_update.each do |name, specifics|
+        # todo each branch
+        uri = specifics[:uri]
+        set_crowdin_specifics(specifics)
+
+        within_tmp_directory(path: File.join(FileUtils.pwd, name), debug: debug) do
+          git_repo = GitRepository.new(uri, FileUtils.pwd)
+          git_repo.clone
+
+          git_repo.checkout(branch)
+          # todo or should we merge this branch into the next ('push vs pull')
+          git_repo.merge(previous_branch, '-Xours') if previous_branch
+          git_repo.within_repo do
+            begin
+              upload_english
+              request_build
+              download_and_replace_locales
+            rescue Crowdin::API::Errors::Error => e
+              puts "Error during update of #{name}: #{e.message}"
+            end
+            fix_missing_keys
+          end
+          git_repo.add('config/locales')
+          git_repo.commit('update locales from crowdin')
+          git_repo.push(branch) unless debug
+        end
+      end
+    end
+  end
+
+  def self.plugins_with_locales
+    configuration[:plugins]
+  end
+
+  def self.configuration
+    # todo raise error if something is missing
+    @configuration ||= begin
+      configuration_file = Pathname(__FILE__) + '../../../../../'
+    configuration_file = configuration_file + 'configuration.yml'
+    YAML.load_file(configuration_file)
+    end
+  end
+
+  def self.set_crowdin_specifics(configuration_hash)
+    @project_id = configuration_hash[:project_id]
+    @api_key = configuration_hash[:api_key]
+    @crowdin_directory = configuration[:crowdin_directory]
+  end
+
+  def self.branch
+    configuration[:branch]
+  end
+
+  def self.previous_branch
+    configuration[:previous_branch]
+  end
+
+  def self.upload_english
+    # create crowdin directory just in case it doesn't exist.
+    crowdin = create_crowdin_handle
+    dir = crowdin.project_info['files'].find {|f| f['name'] == @crowdin_directory && f['node_type'] == 'directory'}
+    unless dir
+      crowdin.add_directory(@crowdin_directory)
+    end
+
+    # either add or update the english (js) translation file
+    titles = {
+      ENGLISH_TRANLATION_FILE => 'OpenProject Wording',
+      ENGLISH_JS_TRANLATION_FILE => 'OpenProject JavaScript Wording'
+    }
+    [ENGLISH_TRANLATION_FILE, ENGLISH_JS_TRANLATION_FILE].each do |translation_file|
+      path_to_translation = File.join 'config', 'locales', translation_file
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.